### PR TITLE
Introduce TokenBuffer

### PR DESCRIPTION
**Pull request by edwin0cheng**
As discussed in Zulip, this PR Introduce `TokenBuffer` , a safe version of  `syn` crate `TokenBuffer` which support cursor based traversal of `tt::TokenTree`. This is the basis of incoming refactoring of `TokenSource` iterator based API.

This PR do the following things:

* Add TokenBuffer in `ra_tt` crate.
* Try to use this new API to refactor the `SubtreeSource` to prove it usage.

### CODE REVIEW

**Comment by matklad** [comment]
By the way, I've recently seen a cool trick for adding "random-access to element"  to a tree structure, in IntelliJ's Rope

The trick is to cache the last accessed leaf node: if the next random index happens to be in the same node (which is common), you don't need to re-fetch the node from the root.

See what happens with this field:

https://github.com/JetBrains/intellij-community/blob/b11987aed9003f4517d1cf6ad84e911852ba7a4e/platform/util/src/com/intellij/util/text/ImmutableText.java#L69

File: crates/ra_tt/src/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        let mut buffers = vec![];
+
+        let idx = TokenBuffer::new_inner(tokens.to_vec(), &mut buffers, None);
+        assert_eq!(idx, 0);
+
+        TokenBuffer { buffers }
+    }
+
+    fn new_inner(
+        tokens: Vec<TokenTree>,
```
Comment (matklad) [inline crates/ra_tt/src/buffer.rs]: Do we really need a Vec here? we are doing `.into_iter`, so perhaps we can passed in `impl Iterator<Item=TokenTree>`? The call site than can do `.cloned()`.

Comment (edwin0cheng) [reply crates/ra_tt/src/buffer.rs]: Fixed !

**Pull request merged**